dotnet/src/SemanticKernel.Abstractions/Contents/BinaryContent.cs

```diff
@@ -3,118 +3,310 @@
+        }
 
-        this._streamProvider = streamProvider;
+        this.MimeType = mimeType;
+        this.Data = data;
     }
 
+    #region Private
     /// <summary>
-    /// Access the content stream.
+    /// Sets the Uri of the content.
     /// </summary>
-    /// <remarks>
-    /// Caller responsible for disposal.
-    /// </remarks>
-    public async Task<Stream> GetStreamAsync()
+    /// <param name="uri">Target Uri</param>
+    private void SetUri(Uri? uri)
     {
-        if (this._streamProvider is not null)
+        if (uri?.Scheme == "data")
         {
-            return await this._streamProvider.Invoke().ConfigureAwait(false);
+            throw new InvalidOperationException("For DataUri contents, use DataUri property.");
         }
 
-        if (this.Content is not null)
+        this._referencedUri = uri;
+    }
+
+    /// <summary>
+    /// Gets the Uri of the content.
+    /// </summary>
+    /// <returns>Uri of the content</returns>
+    private Uri? GetUri()
+        => this._referencedUri;
+
+    /// <summary>
+    /// Sets the DataUri of the content.
+    /// </summary>
+    /// <param name="dataUri">DataUri of the content</param>
+    /// <exception cref="ArgumentException"></exception>
+    private void SetDataUri(string? dataUri)
+    {
+        if (dataUri is null)
         {
-            return new MemoryStream(this.Content.Value.ToArray());
+            this._dataUri = null;
+
+            // Invalidate the current bytearray
+            this._data = null;
+            return;
         }
 
-        throw new KernelException("Null content");
+        var isDataUri = dataUri!.StartsWith("data:", StringComparison.OrdinalIgnoreCase) == true;
+        if (!isDataUri)
+        {
+            throw new UriFormatException("Invalid data uri. Scheme should start with \"data:\"");
+        }
+
+        // Validate the dataUri format
+        var parsedDataUri = DataUriParser.Parse(dataUri);
+
+        // Overwrite the mimetype to the DataUri.
+        this.MimeType = parsedDataUri.MimeType;
+
+        // If parameters where provided in the data uri, updates the content metadata.
+        if (parsedDataUri.Parameters.Count != 0)
+        {
+            this.UpdateDataUriParametersToMetadata(parsedDataUri);
+        }
+
+        this._dataUri = dataUri;
+
+        // Invalidate the current bytearray
+        this._data = null;
+    }
+
+    private void UpdateDataUriParametersToMetadata(DataUriParser.DataUri parsedDataUri)
+    {
+        var newMetadata = this.Metadata as Dictionary<string, object?>;
+        if (newMetadata is null)
+        {
+            // Clone the current read only metadata;
+            newMetadata = new Dictionary<string, object?>();
+            if (this.Metadata is not null)
+            {
+                foreach (var property in this.Metadata)
+                {
+                    newMetadata[$"data-uri-{property.Key}"] = property.Value;
```

Comment:
If I'm reading this right, it seems like this bit of code will only execute if Metadata is not castable to Dictionary<string, object>, and in that case we copy everything that was in the IReadOnlyDictionary<string, object?> to the new dictionary, but we rename each value by prepending `data-uri-`.  I don't understand why we are doing this.  Am I missing something?

Reply:
You are right, good catch there, I added a UT to ensure I don't rename existing non-data-uri parameters and changed the logic on this block.